### PR TITLE
Challenge type and Proposal polymorphism in the API

### PR DESCRIPTION
**Pull request by mzabaluev**
Define different proposal subtypes based on the value of the `challenge_type` field.

### CODE REVIEW

File: doc/api/v0.yaml [hunk 1]
```diff
@@ -227,12 +227,47 @@ components:
         fund_id:
           type: integer
           format: int32
+        challenge_id:
+          type: integer
+          format: int32
+        challenge_type:
+          $ref: "#/components/schemas/ChallengeType"
```
Comment (mzabaluev) [inline doc/api/v0.yaml:234]: Perhaps it's better to group these in an object that shares part of the structure with `Challenge`, also adding the challenge title to facilitate use in an UI view.

Comment (mzabaluev) [reply doc/api/v0.yaml:234]: Or would be better, but then the discriminator thing will not work.

File: doc/api/v0.yaml [hunk 2]
```diff
@@ -51,7 +51,7 @@ paths:
             application/json:
               schema:
                 items:
-                  $ref: "#/components/schemas/Proposal"
+                $ref: "#/components/schemas/ProposalWithChallengeInfo"
```
Comment (eugene-babichenko) [inline doc/api/v0.yaml:54]: Wrong indentation level.

File: doc/api/v0.yaml [hunk 3]
```diff
@@ -224,15 +219,61 @@ components:
...
+        - $ref: "#/components/schemas/ProposalWithChallengeInfo"
+        - type: object
+          properties:
+            proposal_brief:
+              type: string
+            proposal_importance:
+              type: string
+            proposal_goal:
+              type: string
+            proposal_metrics:
```
Comment (eugene-babichenko) [inline doc/api/v0.yaml:267]: Is this also some textual description?

File: doc/api/v0.yaml [hunk 4]
```diff
@@ -224,15 +219,61 @@ components:
...
+      type: string
+      enum:
+        - simple
+        - community-choice
+
+    ProposalWithChallengeInfo:
+      discriminator:
+        propertyName: challenge_type
+        mapping:
+          simple: SimpleProposal
```
Comment (eugene-babichenko) [inline doc/api/v0.yaml:233]: This should be a fully qualified path to the type (like in `$ref`).

**Review by eugene-babichenko — CHANGES_REQUESTED** [review]
We need an automatic OpenAPI validator :)

File: doc/api/v0.yaml (hunk 3 above)
Comment (mzabaluev) [reply doc/api/v0.yaml:299]: Yes, just another field describing the key metrics.
The cue is that all of these fields need proper OpenAPI descriptions.

File: doc/api/v0.yaml (hunk 4 above)
Comment (mzabaluev) [reply doc/api/v0.yaml:233]: Do you think the example [in the spec](http://spec.openapis.org/oas/v3.0.3#discriminator-object) may be misleading?

Comment (eugene-babichenko) [reply doc/api/v0.yaml:233]: It is. The documentation on swagger.io provides an [example with fully-qualified path](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/).

File: doc/api/v0.yaml (hunk 2 above)
Comment (mzabaluev) [reply doc/api/v0.yaml]: Good catch. Both levels are actually somewhat valid, so I guess this is why the extension I have for auto-formatting and linting OpenAPI did not complain.

File: doc/api/v0.yaml (hunk 4 above)
Comment (mzabaluev) [reply doc/api/v0.yaml:233]: I actually think both are valid from the language of the spec, the swagger example just being extra pedantic.

**Comment by mzabaluev** [comment]
> We need an automatic OpenAPI validator :)

Agree. Feel free to add a step to check this in the `api_gh` workflow.

File: .github/workflows/api_validate.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+name: API Check
```
Comment (danielSanchezQ) [inline .github/workflows/api_validate.yml:1]: Looks like there are two jobs that validate the API

Comment (mzabaluev) [reply .github/workflows/api_validate.yml]: Yeah, fixed that in a force-push.

**Review by eugene-babichenko — APPROVED** [review]

**Comment by mzabaluev** [comment]
I got a little carried away with the linter, should be good now.

**Pull request merged**